[mlir][mlir-spirv-cpu-runner] Move MLIR pass pipeline to mlir-opt

andfau-amd opened the pull request:
Adds a new mlir-opt test-only pass, -test-spirv-cpu-runner-pipeline, which runs the set of MLIR passes needed for the mlir-spirv-cpu-runner, and removes them from the runner. The tests are changed to invoke mlir-opt with this flag before running the runner. The eventual goal is to move all host/device code generation steps out of the runner, like with some of the other runners.

andfau-amd commented:
For broader motivation see https://github.com/llvm/llvm-project/issues/73457.

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass for use by mlir-spirv-cpu-runner tests.
```
```suggestion
// Implements a pass for use by mlir-spirv-cpu-runner tests.
```

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+
+namespace {
+
+class TestSPIRVCPURunnerPipelinePass
+    : public PassWrapper<TestSPIRVCPURunnerPipelinePass,
```
```suggestion
struct TestSPIRVCPURunnerPipelinePass final
    : PassWrapper<TestSPIRVCPURunnerPipelinePass,
```

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+class TestSPIRVCPURunnerPipelinePass
+    : public PassWrapper<TestSPIRVCPURunnerPipelinePass,
+                         OperationPass<ModuleOp>> {
+public:
```
```suggestion
```

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+public:
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(TestSPIRVCPURunnerPipelinePass)
+
+  StringRef getArgument() const final {
```
```suggestion
  StringRef getArgument() const override {
```

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+  StringRef getArgument() const final {
+    return "test-spirv-cpu-runner-pipeline";
+  }
+  StringRef getDescription() const final {
```
```suggestion
  StringRef getDescription() const override {
```

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  TestSPIRVCPURunnerPipelinePass() = default;
+  TestSPIRVCPURunnerPipelinePass(const TestSPIRVCPURunnerPipelinePass &) {}
```
Also `= default`? Or should this be deleted?

andfau-amd replied:
Oh I'm not sure. This is an explicit copy-constructor right? I don't know why someone would or wouldn't have one for a pass.

kuhar replied:
I'd make it `= default;` unless we have justification for doing something else. Or remove it if it's not needed.

andfau-amd replied:
I dropped both constructors and it still seems to work. Are you happier with this?

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,87 @@
+};
+} // namespace
+
+namespace mlir {
+namespace test {
+void registerTestSPIRVCPURunnerPipelinePass() {
+  PassRegistration<TestSPIRVCPURunnerPipelinePass>();
+}
+} // namespace test
+} // namespace mlir
```
This should be better at preventing link errors.

```suggestion
void mlir::test::registerTestSPIRVCPURunnerPipelinePass() {
  PassRegistration<TestSPIRVCPURunnerPipelinePass>();
}
```

kuhar replied:
Ah never mind, I see why we need to have it this way -- it's not declared in any header

andfau-amd replied:
It would, but unfortunately C++ would require us to still have that ugly double namespace for a forward declaration, otherwise:

```
mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:80:12: error: no member named 'test' in namespace 'mlir'
void mlir::test::registerTestSPIRVCPURunnerPipelinePass() {
     ~~~~~~^
1 error generated.
ninja: build stopped: subcommand failed.
```

kuhar reviewed (commented):
Nice!

kuhar commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,83 @@
+
+namespace {
+
+struct TestSPIRVCPURunnerPipelinePass
+    : public PassWrapper<TestSPIRVCPURunnerPipelinePass,
```
```suggestion
struct TestSPIRVCPURunnerPipelinePass final
    : PassWrapper<TestSPIRVCPURunnerPipelinePass,
```

kuhar reviewed (approved):
LGTM % nit

kuhar reviewed (approved):
Thanks for the fixes

joker-eph commented on mlir/test/lib/Pass/TestSPIRVCPURunnerPipeline.cpp:
```diff
@@ -0,0 +1,82 @@
+    if (failed(runPipeline(passManager, module)))
+      signalPassFailure();
+  }
+};
```
Wondering whether this could be registered as a pipeline instead of a pass?

kuhar replied:
Do we have support for **test pipeline** registration? The only other one I saw was also implemented as a pass: https://github.com/llvm/llvm-project/blob/9f3c55954eaa71910caa0abbb404db02d3a104c1/mlir/test/lib/Pass/TestDynamicPipeline.cpp#L21

andfau-amd replied:
That one is what I used as a reference when implementing this, FWIW.

andfau-amd replied:
Thank you very much for the suggestion; thanks to a colleague I was made aware of this example https://github.com/llvm/llvm-project/blob/665457815f11118f7e755a471f33606c8562a4be/mlir/test/lib/Dialect/LLVM/TestLowerToLLVM.cpp#L87 and I was able to rework this into a pipeline quite simply. :)

andfau-amd replied:
Biggest benefit is being able to do away with the explicit dialect registration.

andfau-amd committed:
[mlir][mlir-spirv-cpu-runner] Move MLIR pass pipeline to mlir-opt

Adds a new mlir-opt test-only pipeline, -test-spirv-cpu-runner-pipeline,
containing the set of MLIR passes needed for the mlir-spirv-cpu-runner,
and removes them from the runner. The tests are changed to invoke
mlir-opt with this flag before running the runner. The eventual goal
is to move all host/device code generation steps out of the runner, like
with some of the other runners.

kuhar reviewed (approved):
LGTM

kuhar commented:
Thanks for the suggestion @joker-eph
